enhancements/aws-pod-identity.md

```diff
@@ -0,0 +1,442 @@
+---
+title: aws-pod-identity 
+authors:
+  - "@sjenning"
+reviewers:
+  - "@derekwaynecarr"
+  - "@smarterclayton"
+approvers:
+  - "@derekwaynecarr"
+creation-date: 2020-03-26
+last-updated: 2020-03-26
+status: provisional
+see-also:
+  - ""  
+replaces:
+  - ""
+superseded-by:
+  - ""
+---
+
+# Support Native (IAM-based) Pod Identity on AWS
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+## Summary
+
+When deployed on AWS, Kubernetes can use pod identity managed by IAM.  AWS has a
+[webhook](https://github.com/aws/amazon-eks-pod-identity-webhook/) they maintain
+for EKS that does enables this ability.  This enhancement integrates the same
+functionality into OpenShift.
+
+The `cloud-credential-operator` will be responsible for managing this new
+ability as it serves as an alternative credential _injection_ mechanism versus
+the existing mechanism of injecting an IAM user acccess ID and key into a
+secret.  The existing role/user _creation_ via a CredentialRequest CR is common
+between the methods.
+
+The new injection mechanism works by creating a `ServiceAccount` annotated with
+a IAM role the pod should assume. When a pod runs with the annoated
+`ServiceAccount`, the webhook injects a projected volume into the pod that
+contanis the bound token that can be used to assume the IAM role.  The AWS SDK
+searches for this token automatically.
+
+## Motivation
+
+This allows users to more cleanly express and manage pod identities in IAM when
+running on AWS.  The use of bound tokens to assume roles (vs user id/key) is
+preferred as these tokens can expire and be rotated.
+
+### Goals
+
+1. A pod can assume an IAM role created with a `CredentialRequest` CR using a
+   bounded token
+2. An admin can create a IAM role via `CredentialRequest` CR and inject to a
+   `ServiceAccount` (just as they can create an IAM user and inject to a
+   `Secret` currently)
+3. OpenShift should create the OIDC provider in IAM
+4. The token signing CA (`bound-sa-token-signing-certs`) can rotate and should
+   sync between OCP and the OIDC provider in IAM.  Rotation should be seemsless.
+   Assumes applications in pods monitor the token for rotation (AWS SDK does
+   this)
+
+### Non-Goals
+
+1. Convert existing IAM user-based components (e.g. ingress, image-registry,
```

Comment:
Understood that converting existing components is a non-goal, but can we give an example of what that future flow would look like? Specifically, I'm thinking about an example with some details covering (for instance) this pod identity feature landing in 4.6, and image-registry, ingress, etc aren't converted, but then from 4.6 to 4.7 what does an in-place upgrade flow look like for image-registry, ingress, etc?

Reply:
I think that, for existing clusters, you wouldn't want to change away from the secret injection mode on upgrade.  That hand-off could be tricky and would likely involve code change in the operators that create the `CredentialsRequest` CRs.

We might want to document a way that customers could convert manually though.

Reply:
I looked at some operators and the operators don't dynamically create the CredentialsRequest CRs; they just exist statically in their manifests dir.

This feels like we're leaving all the cluster components stuck with secret injection for pre and post-pod-identity-OpenShift clusters unless we can architect some secret-to-serviceAccount migration path, or perhaps some way to have the operators handle having CredentialsRequests for both secret and serviceAccount methods to support clusters installed on a pre-pod-identity release.

Reply:
@joelddiaz yes, i'm thinking that is a non-goal for this part and that a migration from secret to serviceaccount based for already processed `CredentialsRequest` CRs would need to be done manually at this point in time (i.e. docs).